File: c++projects/LinkedListOperations/ListOps.cpp
#include <iostream>

struct Node
{
    int data;
    Node *next;

    Node(int value) : data(value), next(nullptr) {}
};

class LinkedList
{
private:
    Node *head;

public:
    LinkedList() : head(nullptr) {}

    void insertNode(int data, int position)
    {
        Node *newNode = new Node(data);

        if (position == 0)
        {
            newNode->next = head;
            head = newNode;
            return;
        }

        Node *temp = head;
        for (int i = 0; temp != nullptr && i < position - 1; ++i)
        {
            temp = temp->next;
        }

        if (temp == nullptr)
        {
            std::cout << "Position is out of bounds" << std::endl;
            delete newNode;
            return;
        }

        newNode->next = temp->next;
        temp->next = newNode;
    }

    void deleteNode(int data)
    {
        Node *temp = head;
        Node *prev = nullptr;

        if (temp != nullptr && temp->data == data)
        {
            head = temp->next;
            delete temp;
            return;
        }

        while (temp != nullptr && temp->data != data)
        {
            prev = temp;
            temp = temp->next;
        }

        if (temp == nullptr)
        {
            std::cout << "Node with data " << data << " not found" << std::endl;
            return;
        }

        prev->next = temp->next;
        delete temp;
    }

    void deleteList()
    {
        Node *current = head;
        Node *next = nullptr;

        while (current != nullptr)
        {
            next = current->next;
            delete current;
            current = next;
        }

        head = nullptr;
    }

    void printList() const
    {
        Node *temp = head;
        while (temp != nullptr)
        {
            std::cout << temp->data << " -> ";
            temp = temp->next;
        }
        std::cout << "NULL" << std::endl;
    }

    ~LinkedList()
    {
        deleteList();
    }
};

int main()
{
    LinkedList list;

    list.insertNode(10, 0);
    list.insertNode(20, 1);
    list.insertNode(30, 2);
    list.insertNode(15, 1);

    std::cout << "Linked list after insertions: ";
    list.printList();

    list.deleteNode(20);
    std::cout << "Linked list after deleting node with data 20: ";
    list.printList();

    list.deleteList();
    std::cout << "Linked list after deleting the entire list: ";
    list.printList();

    return 0;
}

File: c++projects/LinkedListOperations/ListOps.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

class LinkedList:
    def __init__(self):
        self.head = None

    def insert_node(self, data, position):
        new_node = Node(data)
        
        if position == 0:
            new_node.next = self.head
            self.head = new_node
            return
        
        temp = self.head
        for _ in range(position - 1):
            if temp is None:
                print("Position is out of bounds")
                return
            temp = temp.next
        
        if temp is None:
            print("Position is out of bounds")
            return
        
        new_node.next = temp.next
        temp.next = new_node

    def delete_node(self, data):
        temp = self.head
        prev = None

        if temp is not None and temp.data == data:
            self.head = temp.next
            return

        while temp is not None and temp.data != data:
            prev = temp
            temp = temp.next

        if temp is None:
            print(f"Node with data {data} not found")
            return

        prev.next = temp.next

    def delete_list(self):
        self.head = None

    def print_list(self):
        temp = self.head
        while temp:
            print(temp.data, end=" -> ")
            temp = temp.next
        print("NULL")

if __name__ == "__main__":
    list = LinkedList()
    
    list.insert_node(10, 0)
    list.insert_node(20, 1)
    list.insert_node(30, 2)
    list.insert_node(15, 1)
    
    print("Linked list after insertions:")
    list.print_list()
    
    list.delete_node(20)
    print("Linked list after deleting node with data 20:")
    list.print_list()
    
    list.delete_list()
    print("Linked list after deleting the entire list:")
    list.print_list()
